[17.09] Fail job if tools that use galaxy.json write to stderr

mvdbeek opened the pull request:
This should fix #5214.

mvdbeek committed:
Read in stderr form galaxy.json

before checking whether stderr and exit code combination should result in a
failed job. This should fix https://github.com/galaxyproject/galaxy/issues/5214.

mvdbeek committed:
Add API tests for history uploads from URL

mvdbeek committed:
Restrict reading tool provided metadata to tools that actually provide metadata

mvdbeek commented:
This integration test fails now: https://jenkins.galaxyproject.org/job/docker-integration/3330/testReport/integration.test_upload_configuration_options/DefaultBinaryContentFiltersTestCase/test_gzipped_html_content_blocked_by_default/ ... I guess it expects a dataset in the OK state, but I'd argue failing it is better, right ?

mvdbeek committed:
Do not assert dataset state in test_gzipped_html_content_blocked_by_default test

jmchilton commented on test/api/test_tools.py:
```diff
@@ -1407,6 +1416,19 @@ def _run(self, tool_id, history_id, inputs, assert_ok=False, tool_version=None):
         else:
             return create_response
 
+    def _upload_from_url(self, url):
```
In this same directory there is a test_tools_upload.py test case filled with tons of upload tests - can you move these there?

mvdbeek replied:
Hmm, that's only in dev. Should I create the file ?

jmchilton replied:
Doh - good point. Lets keep it as is - I'll try to remember to move things around after we merge this forward.

jmchilton commented on test/integration/test_upload_configuration_options.py:
```diff
@@ -82,9 +82,9 @@ def test_random_binary_allowed(self):
+            self.history_id, 'file://%s/bad.html.gz' % TEST_DATA_DIRECTORY, file_type="auto", wait=True, assert_ok=False
         )
-        dataset = self.dataset_populator.get_history_dataset_details(self.history_id, dataset=dataset)
+        dataset = self.dataset_populator.get_history_dataset_details(self.history_id, dataset=dataset, assert_ok=False)
```
Yes - this is good. Thanks!

mvdbeek committed:
Adjust stderr collection to (Legacy-)ToolProvidedMetadata

mvdbeek committed:
Optionally allow dataset populators to fail

jmchilton committed:
Let ToolProvidedMetadata interface more directly decide if it has failed outputs.

I like this better for three reasons:

- Since usually it is scripts producing this JSON - we have the most control at that point for determining the failure and we don't have to deal with an artificial dependency between the tool's stdio and the output.
- At some point we could potentially allow some datasets to be ok now even though the job fails.
- It is a cleaner interface at the Python level between job finish and output collection IMO (no need for isinstance checking).

jmchilton commented:
Any interest in rearranging this so the tool directly just declares the output in question is in error? (https://github.com/mvdbeek/galaxy/pull/6).

mvdbeek committed:
Merge pull request #6 from jmchilton/fail_if_galaxy_json_tools_write_to_stderr

Let ToolProvidedMetadata interface more directly decide if it has failed outputs.

nsoranzo commented on test/api/test_tools.py:
```diff
@@ -1366,6 +1366,15 @@ def test_combined_mapping_and_subcollection_mapping(self):
 
+    def test_upload_from_invalid_url(self):
+        history_id, dataset_id = self._upload_from_url('https://usegalaxy.org/bla123')
+        datset_details = self.dataset_populator.get_history_dataset_details(history_id, dataset_id=dataset_id, assert_ok=False)
```
s/datset_details/dataset_details/ 

Also on the line below.

mvdbeek committed:
Fix dataset spelling (thx @nsoranzo)
